Windows Cmake RelWithDebInfo not working

Summary:
I'm having an issue getting TensorFlow to build with debug symbols in Windows using Cmake.

About my system:

Windows 10
Version 1607
Build 14393.693

Cmake version 3.7.1

Visual Studio Community 2015

Clone tensor flow repository from https://github.com/tensorflow/tensorflow

git clone https://github.com/tensorflow/tensorflow.git

Change directory to tensorflow\tensorflow\contrib\cmake

cd gitFolder\tensorflow\tensorflow\contrib\cmake


Make build directory, change to directory

Mkdir build

cd build

Use cmake

Cmake .. -A  x64 -DCMAKE_BUILD_TYPE=RelWithDebInfo ^
-DSWIG_EXECUTABLE=C:\tools\swigwin-3.0.10\swig.exe ^
-DPYTHON_EXECUTABLE=C:\Users\ian\Anaconda3\python.exe ^
-DPYTHON_LIBRARIES=C:\Users\ian\Anaconda3\python35.lib ^
-Dtensorflow_ENABLE_GRPC_SUPPORT=OFF

Note:
Swig, python executable directory and python library directory will change based on your system. I tried at with the GRPC support enabled (this is the default) but I got build errors related to GRPC. Because we will be handling remote web requests with our own server application, I didn’t think we needed the GRPC support.

Open the solution in Visual Studio.

Change build mode to RelWithDebInfo x64

Build

Result:

Most projects build, but sadly not the one that I’m interested in, tf_label_image_example. The errors I received are:
Link1120 1 Unresolved externals pywrap_tensorflow
Link1318  Unexpected PDB error; OK(0) tf_tutorials_example_trainer
Link1318  Unexpected PDB error; OK(0) tf_label_image_example
LNK2019	unresolved external symbol "class tensorflow::Status __cdecl tensorflow::NewServer(class tensorflow::ServerDef const &,class std::unique_ptr<class tensorflow::ServerInterface,struct std::default_delete<class tensorflow::ServerInterface> > *)" (?NewServer@tensorflow@@YA?AVStatus@1@AEBVServerDef@1@PEAV?$unique_ptr@VServerInterface@tensorflow@@U?$default_delete@VServerInterface@tensorflow@@@std@@@std@@@Z) referenced in function "void __cdecl PyServer_New(class tensorflow::ServerDef const &,class std::unique_ptr<class tensorflow::ServerInterface,struct std::default_delete<class tensorflow::ServerInterface> > *,struct TF_Status *)" (?PyServer_New@@YAXAEBVServerDef@tensorflow@@PEAV?$unique_ptr@VServerInterface@tensorflow@@U?$default_delete@VServerInterface@tensorflow@@@std@@@std@@PEAUTF_Status@@@Z)	pywrap_tensorflow	C:\tesorflow\tensorflow\tensorflow\contrib\cmake\build\pywrap_tensorflow.obj

OK, it fails to build the pywrap DLL because it failed to compile the wrapper. And that's because it failed to resolve the C++ function. Could you find a way to get the command-line that's running on this? It should be trying to link against the C++ tensorflow lib. We'd need to check the symbols in the C++ tensorflow lib and find out what symbols are out there. Maybe they're x86 symbols, or they have the wrong linkage.

```
194>pywrap_tensorflow.obj : error LNK2019: unresolved external symbol "class tensorflow::Status __cdecl tensorflow::NewServer(class tensorflow::ServerDef const &,class std::unique_ptr<class tensorflow::ServerInterface,struct std::default_delete<class tensorflow::ServerInterface> > *)" (?NewServer@tensorflow@@YA?AVStatus@1@AEBVServerDef@1@PEAV?$unique_ptr@VServerInterface@tensorflow@@U?$default_delete@VServerInterface@tensorflow@@@std@@@std@@@Z) referenced in function "void __cdecl PyServer_New(class tensorflow::ServerDef const &,class std::unique_ptr<class tensorflow::ServerInterface,struct std::default_delete<class tensorflow::ServerInterface> > *,struct TF_Status *)" (?PyServer_New@@YAXAEBVServerDef@tensorflow@@PEAV?$unique_ptr@VServerInterface@tensorflow@@U?$default_delete@VServerI
```

Just above there's 

```
196>utils.obj : fatal error LNK1318: Unexpected PDB error; OK (0) ''
195>utils.obj : fatal error LNK1318: Unexpected PDB error; OK (0) ''
```

--- a/tensorflow/contrib/cmake/tf_core_cpu.cmake
+++ b/tensorflow/contrib/cmake/tf_core_cpu.cmake
@@ -6,6 +6,8 @@ file(GLOB_RECURSE tf_core_cpu_srcs
     "${tensorflow_source_dir}/tensorflow/cc/saved_model/*.cc"
     "${tensorflow_source_dir}/tensorflow/core/common_runtime/*.h"
     "${tensorflow_source_dir}/tensorflow/core/common_runtime/*.cc"
+    "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/server_lib.h"
+    "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/server_lib.cc"
     "${tensorflow_source_dir}/tensorflow/core/graph/*.h"
     "${tensorflow_source_dir}/tensorflow/core/graph/*.cc"
     "${tensorflow_source_dir}/tensorflow/core/public/*.h"

--- a/tensorflow/contrib/cmake/tf_core_distributed_runtime.cmake
+++ b/tensorflow/contrib/cmake/tf_core_distributed_runtime.cmake
@@ -7,6 +7,7 @@ file(GLOB_RECURSE tf_core_distributed_runtime_srcs
 )
 
 file(GLOB_RECURSE tf_core_distributed_runtime_exclude_srcs
+    "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/server_lib.cc"  # Build in tf_core_cpu instead.
     "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/*test*.h"
     "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/*test*.cc"
     "${tensorflow_source_dir}/tensorflow/core/distributed_runtime/rpc/grpc_tensorflow_server.cc"

--- a/tensorflow/contrib/cmake/tf_tools.cmake
+++ b/tensorflow/contrib/cmake/tf_tools.cmake
@@ -18,10 +18,10 @@ target_link_libraries(${proto_text} PUBLIC
   tf_protos_cc
 )
 
-add_dependencies(${proto_text}
-    tf_core_lib
-    grpc
-)
+add_dependencies(${proto_text} tf_core_lib)
+if(tensorflow_ENABLE_GRPC_SUPPORT)
+    add_dependencies(${proto_text} grpc)
+endif(tensorflow_ENABLE_GRPC_SUPPORT)
 
 file(GLOB_RECURSE tf_tools_transform_graph_lib_srcs
     "${tensorflow_source_dir}/tensorflow/tools/graph_transforms/*.h"

